File: 930.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        int result = 0;
        int prefixSum = 0;

        for(int num : nums) {
            prefixSum += num;
            int key = prefixSum - goal;
            if(map.containsKey(key)){
                result += map.get(key);
            }
            map.put(prefixSum, map.getOrDefault(prefixSum, 0) + 1);
        }
        return result;
    }
}

File: 930.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        count = defaultdict(int)
        count[0] = 1
        prefixsum = 0
        result = 0

        for num in nums:
            prefixsum += num
            result += count[prefixsum - goal]
            count[prefixsum] += 1
        
        return result
